Source/WebCore/css/parser/CSSPropertyParser.cpp

```diff
@@ -2332,29 +2333,37 @@ bool CSSPropertyParser::consumeGridShorthand(bool important)
 
 bool CSSPropertyParser::consumePlaceContentShorthand(bool important)
 {
+    // <'place-content'>   = <'align-content'> <'justify-content'>?
+    // <'align-content'>   = normal | <baseline-position> | <content-distribution> | <overflow-position>? [ <content-position>                ]
+    // <'justify-content'> = normal |                       <content-distribution> | <overflow-position>? [ <content-position> | left | right ]
+    // https://drafts.csswg.org/css-align/#propdef-place-content
+
+    // NOTE: The second value, if omitted, is copied from the first value.
+
     ASSERT(shorthandForProperty(CSSPropertyPlaceContent).length() == 2);
 
     if (m_range.atEnd())
         return false;
 
-    CSSParserTokenRange rangeCopy = m_range;
     bool isBaseline = isBaselineKeyword(m_range.peek().id());
-    RefPtr alignContentValue = consumeContentDistributionOverflowPosition(m_range, isContentPositionKeyword);
+
+    RefPtr alignContentValue = consumeAlignContent(m_range);
     if (!alignContentValue)
         return false;
 
-    // justify-content property does not allow the <baseline-position> values.
-    if (m_range.atEnd() && isBaseline)
-        return false;
-    if (isBaselineKeyword(m_range.peek().id()))
-        return false;
+    if (m_range.atEnd()) {
+        // justify-content property does not allow the <baseline-position> values, so if only 'align-content' is provided, it can't be a <baseline-position>.
+        if (isBaseline)
+            return false;
 
-    if (m_range.atEnd())
-        m_range = rangeCopy;
-    RefPtr justifyContentValue = consumeContentDistributionOverflowPosition(m_range, isContentPositionOrLeftOrRightKeyword);
-    if (!justifyContentValue)
-        return false;
-    if (!m_range.atEnd())
+        RefPtr justifyContentValue = alignContentValue.copyRef();
+        addProperty(CSSPropertyAlignContent, CSSPropertyPlaceContent, alignContentValue.releaseNonNull(), important);
```

Comment:
Maybe we should add a helper that sets two properties to the same value to clean up the idiom in all of these clauses. Since there are 4 of them that seems like enough to justify it.

Reply:
I think that is a great idea! I have a few questions about how to get it right:

1. Is consume2ValueShorthand() already doing this? (as much time as I have spent on property parsing, I have spent little on the shorthands).

2. If we use the pattern I used exactly, we could have something like:

```c++
bool CSSPropertyParser::consumePairedShorthand(const StylePropertyShorthand& shorthand, bool important)
{
    ASSERT(shorthand.length() == 2);
    const CSSPropertyID* longhands = shorthand.properties();

    RefPtr prop1 = parseSingleValue(longhands[0], shorthand.id());
    if (!prop1)
        return false;

    if (m_range.atEnd()) {
        RefPtr prop2 = prop1;
        addProperty(longhands[0], shorthand.id(), prop1.releaseNonNull(), important);
        addProperty(longhands[1], shorthand.id(), prop2.releaseNonNull(), important);
        return true;
    }

    RefPtr prop2 = parseSingleValue(longhands[1], shorthand.id());
    if (!prop2 || !m_range.atEnd())
        return false;

    addProperty(longhands[0], shorthand.id(), prop1.releaseNonNull(), important);
    addProperty(longhands[1], shorthand.id(), prop2.releaseNonNull(), important);
    return true;
}
```

But that would leave out `place-content` since that has an extra check.

Alternatively, we could back to the way they were doing it like:

```c++
bool CSSPropertyParser::consumePairedShorthand(const StylePropertyShorthand& shorthand, bool important)
{
    ASSERT(shorthand.length() == 2);
    const CSSPropertyID* longhands = shorthand.properties();

    auto rangeCopy = m_range;

    RefPtr prop1 = parseSingleValue(longhands[0], shorthand.id());
    if (!prop1)
        return false;

    // If there are no more tokens, that prop2 should use re-use the original range. This is the equivalent of copying and validating prop1.
    if (m_range.atEnd())
        m_range = rangeCopy;

    RefPtr prop2 = parseSingleValue(longhands[1], shorthand.id());
    if (!prop2 || !m_range.atEnd())
        return false;

    addProperty(longhands[0], shorthand.id(), prop1.releaseNonNull(), important);
    addProperty(longhands[1], shorthand.id(), prop2.releaseNonNull(), important);
    return true;
}
```

3. Have a good name for this? (`consumePairedShorthand` was a placeholder).

Reply:
I like what you ended up doing. Sorry I didn’t help more with the exploration of possibilities!